crates/extensions/tedge_health_ext/src/lib.rs

```diff
@@ -13,50 +13,80 @@ use tedge_actors::RuntimeRequestSink;
 use tedge_actors::ServiceConsumer;
 use tedge_actors::ServiceProvider;
 use tedge_actors::SimpleMessageBoxBuilder;
-use tedge_api::health::health_status_down_message;
-use tedge_api::health::health_status_up_message;
+use tedge_api::entity_store::EntityRegistrationMessage;
+use tedge_api::entity_store::EntityType;
+use tedge_api::health::ServiceHealthTopic;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::mqtt_topics::Service;
+use tedge_mqtt_ext::Message;
 use tedge_mqtt_ext::MqttConfig;
 use tedge_mqtt_ext::MqttMessage;
 use tedge_mqtt_ext::TopicFilter;
 
 pub struct HealthMonitorBuilder {
-    service_name: String,
+    registration_message: Option<Message>,
+    health_topic: ServiceHealthTopic,
     box_builder: SimpleMessageBoxBuilder<MqttMessage, MqttMessage>,
 }
 
 impl HealthMonitorBuilder {
-    pub fn new(
-        service_name: &str,
+    /// Creates a HealthMonitorBuilder that creates a HealthMonitorActor with
+    /// a new topic scheme.
+    pub fn from_service_topic_id(
+        service: Service,
         mqtt: &mut (impl ServiceProvider<MqttMessage, MqttMessage, TopicFilter> + AsMut<MqttConfig>),
+        // TODO: pass it less annoying way
+        mqtt_schema: &MqttSchema,
+        service_type: String,
     ) -> Self {
-        // Connect this actor to MQTT
+        let service_topic_id = &service.service_topic_id;
+
+        let mut box_builder = SimpleMessageBoxBuilder::new(service_topic_id.as_str(), 16);
+
+        let service_name = service_topic_id.entity().default_service_name().unwrap();
         let subscriptions = vec![
             "tedge/health-check",
             &format!("tedge/health-check/{service_name}"),
         ]
         .try_into()
         .expect("Failed to create the HealthMonitorActor topic filter");
 
-        let mut box_builder = SimpleMessageBoxBuilder::new(service_name, 16);
         box_builder
             .set_request_sender(mqtt.connect_consumer(subscriptions, box_builder.get_sender()));
 
+        let registration_message = EntityRegistrationMessage {
+            topic_id: service_topic_id.entity().clone(),
+            external_id: None,
+            r#type: EntityType::Service,
+            parent: Some(service.device_topic_id.entity().clone()),
+            other: serde_json::json!({"type": service_type}),
+        };
+        let registration_message = registration_message.to_mqtt_message(mqtt_schema);
+
+        let health_topic = ServiceHealthTopic::from_new_topic(service_topic_id, mqtt_schema);
+
         let builder = HealthMonitorBuilder {
-            service_name: service_name.to_owned(),
+            health_topic,
+            registration_message: Some(registration_message),
             box_builder,
         };
 
         // Update the MQTT config
+
+        // XXX: if the same MqttActorBuilder is used in different actors, then
+        // this will override init messages that may have been set by other
+        // actors!
```

Comment:
Only a single instance of the `HealthMonitor` actor should be built.

Reply:
I didn't mean another `HealthMonitor`, but potentially another actor. I wanted to highlight that if multiple actors need to communicate with a single MQTT Actor, then exposing this MQTT actor's internal configuration for other actors so that they can call builder methods on MQTT Actor config, they can overwrite each other's i.e. initial messages. I tried to use `mqtt_channel::Config::with_initial_message` in `MqttBuilderActor`, but it was overridden by `.with_initial_message` call in `HealthMonitorBuilder`.

Reply:
It's true that sharing an MQTT connection with several actors raises questions (session name, last will message). This is why the mapper has two MQTT connection: a main one to send/receive messages and a second one used only for a last will to be forwarded to c8y (see https://github.com/thin-edge/thin-edge.io/pull/2276#discussion_r1360691979).

That said, there is no reasons to forbid independent actors to add their own init messages. For that, `mqtt_channel::Config` must be updated to handle a `Vec<InitMessageFn>` instead of an `Option<InitMessageFn>`.